ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,104 @@ let host_supports_hvm ~__context host =
 let env_with_path env_vars =
   Array.of_list @@
   List.map (fun (k,v) -> Printf.sprintf "%s=%s" k v) (("PATH", String.concat ":" Forkhelpers.default_path)::env_vars)
+
+(** wrapper around the stunnel@xapi systemd service.
+  * there exist scripts (e.g. xe-toolstack-restart) which also manipulate
+    the stunnel daemon but they do this directly (not via ocaml). *)
+module Stunnel : sig
+  (** restart stunnel, possibly changing the config file *)
+  val restart : accept:string -> unit
+end = struct
+
+  let cert = !Xapi_globs.server_cert_path
+
+  (** protect ourselves from concurrent writes to files *)
+  module File : sig
+    (** create or update stunnel config file *)
+    val update_config : accept:string -> unit
+    (** calls `generate_ssl_cert` bash script
+      * TODO: don't generate certificates in bash? *)
+    val gen_pem_if_not_exist : unit -> unit
+  end = struct
+    let xapi_stunnel_conf = "/etc/stunnel/xapi.conf"
+    let m = Mutex.create ()
+    let current_accept = ref None
+
+    let rewrite_xapi_ssl_config_file ~accept =
+      let idle_timeout = match Sys.getenv_opt "STUNNEL_IDLE_TIMEOUT" with
+                         | None -> "; no idle timeout"
+                         | Some x -> Printf.sprintf "TIMEOUTidle = %s" x
+      in
+      let conf_contents = String.concat "\n" [
+          "; autogenerated by helpers.ml"
+        ; "fips = no"
+        ; "pid = /var/run/xapissl.pid"
+        ; "socket = r:TCP_NODELAY=1"
+        ; "socket = a:TCP_NODELAY=1"
+        ; "socket = l:TCP_NODELAY=1"
+        ; "socket = r:SO_KEEPALIVE=1"
+        ; "socket = a:SO_KEEPALIVE=1"
+        ; idle_timeout
+        ; "debug = authpriv.5"
+        ; ""
+        ; "[xapi]"
+        ; Printf.sprintf "accept = %s" accept
+        ; "connect = 80"
+        ; Printf.sprintf "cert = %s" cert
+        ; Printf.sprintf "ciphers = %s" Xcp_const.good_ciphersuites
+        ; "curve = secp384r1"
+        ; "TIMEOUTclose = 1"
+        ; "options = CIPHER_SERVER_PREFERENCE"
+        ; "sslVersion = TLSv1.2"
+        ]
+      in
+      if Sys.file_exists xapi_stunnel_conf then Sys.remove xapi_stunnel_conf;
+      Stdext.Unixext.write_string_to_file xapi_stunnel_conf conf_contents
```

Comment:
There is a function that does an atomic replace, we could use that, then you don't need to remove the old one, the atomic replace will take care of it.